src/peft/utils/other.py

```diff
@@ -481,3 +481,116 @@ def id_tensor_storage(tensor: torch.Tensor) -> Tuple[torch.device, int, int]:
+    if input_ids is not None:
+        input_id_paddings = [input_ids[i][:first_ind] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+        input_ids = [input_ids[i][first_ind:] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+    else:
+        input_embed_paddings = [inputs_embeds[i][:first_ind] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+        inputs_embeds = [inputs_embeds[i][first_ind:] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+    # get the same slices for attention_mask and labels. Note that all the padding vectors are of different shapes
+    # so we can't stack them into a tensor
+    attention_mask_paddings, labels_paddings = None, None
+    if attention_mask is not None:
+        attention_mask_paddings = [
+            attention_mask[i][:first_ind] for i, first_ind in enumerate(first_non_padding_tok_ind)
+        ]
+        attention_mask = [attention_mask[i][first_ind:] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+
+    if labels is not None:
+        labels_paddings = [labels[i][:first_ind] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+        labels = [labels[i][first_ind:] for i, first_ind in enumerate(first_non_padding_tok_ind)]
+
+    pad_els = (input_id_paddings, input_embed_paddings, attention_mask_paddings, labels_paddings)
+
+    return input_ids, inputs_embeds, attention_mask, labels, pad_els
+
+
+def add_pad_tokens(
+    inputs_embeds: Union[List[torch.Tensor], torch.Tensor],
+    attention_mask: Union[List[torch.Tensor], torch.Tensor],
+    labels: Union[List[torch.Tensor], torch.Tensor],
+    pad_els: Optional[Tuple[List[torch.Tensor], List[torch.Tensor], List[torch.Tensor]]],
+):
+    """
+    Adds the padding token elements back to inputs_embeds, attention_mask and labels.
+
+    Returns:
+    - inputs_embeds: inputs_embeds with padding tokens added
+    - attention_mask: attention_mask with padding tokens added
+    - labels: labels with padding tokens added
+    """
+    if pad_els is None:
+        return inputs_embeds, attention_mask, labels  # no special handling needed
+
+    pad_input_ids, pad_inputs_embeds, pad_mask, pad_labels = pad_els
+    final_inputs_embeds = torch.stack(
+        [torch.cat((pad_embed, input_embed)) for input_embed, pad_embed in zip(inputs_embeds, pad_inputs_embeds)]
+    )
+    final_attention_mask, final_labels = None, None
+    if attention_mask is not None:
+        final_attention_mask = torch.stack(
+            [torch.cat((pad_mask, attention_mask)) for attention_mask, pad_mask in zip(attention_mask, pad_mask)]
+        )
+    if labels is not None:
+        final_labels = torch.stack([torch.cat((pad_label, label)) for label, pad_label in zip(labels, pad_labels)])
+    return final_inputs_embeds, final_attention_mask, final_labels
+
+
+def apply_to_list_or_tensor(f, x: Union[List[torch.Tensor], torch.Tensor]) -> Union[List[torch.Tensor], torch.Tensor]:
```

Comment:
Nit: The "more correct" type annotation should be:

```python
from typing import overload

@overload
def apply_to_list_or_tensor(f, x: List[torch.Tensor]) -> List[torch.Tensor]:
   ...

def apply_to_list_or_tensor(f, x: torch.Tensor) -> torch.Tensor:
   # actual code
```

But we're not really strict with type annotations, so it's okay :)